8282395: URL.openConnection can throw IOOBE

tkiriyama opened the pull request:
I fixed sun.net.www.ParseUtil.decode().

ParseUtil.decode() always tries to decode after parsing '%', so if '%' is located at the end of the String, IndexOutOfBoundsException is thrown. Also, if '%' is shown after decodable string and following string is not decodable (e.g: "%25%s%G1"), ParseUtil.decode() throws IllegalArgumentException.

But URL standard says below (https://url.spec.whatwg.org/#percent-decode).

```
Otherwise, if byte is 0x25 (%) and the next two bytes after byte in input are not in the ranges 
0x30 (0) to 0x39 (9), 0x41 (A) to 0x46 (F), and 0x61 (a) to 0x66 (f), all inclusive, append byte to output.
```

So, there should be used isEscaped() to judge to decode.

Would you please review this fix?

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8282395](https://bugs.openjdk.org/browse/JDK-8282395): URL.openConnection can throw IOOBE


### Reviewers
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/8155/head:pull/8155` \
`$ git checkout pull/8155`

Update a local copy of the PR: \
`$ git checkout pull/8155` \
`$ git pull https://git.openjdk.org/jdk pull/8155/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8155`

View PR using the GUI difftool: \
`$ git pr show -t 8155`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/8155.diff">https://git.openjdk.org/jdk/pull/8155.diff</a>

</details>

AlanBateman commented:
Are there examples using URL/URLconnection (rather than ParseUtil directly) to demonstrate the issue?

dfuch commented:
MalformedURLException should probably be thrown instead of simply accepting a malformed % encoded pair.

tkiriyama commented:
> Are there examples using URL/URLconnection (rather than ParseUtil directly) to demonstrate the issue?

I’m sorry for the late reply. Thre is the example in  [JDK-8282395](https://bugs.openjdk.java.net/browse/JDK-8282395).  
This is a simple example code.

```
import java.net.URL;

public class TestURL {
  public static void main(String args[]) throws Exception {
    URL url = new URL("ftp://.:%@");
    url.openConnection();
  }
}
```

The following exception is thrown.

```
>java TestURL
Exception in thread "main" java.lang.IndexOutOfBoundsException: Range [1, 3) out of bounds for length 1
        at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:100)
        at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckFromToIndex(Preconditions.java:112)
        at java.base/jdk.internal.util.Preconditions.checkFromToIndex(Preconditions.java:349)
        at java.base/java.util.Objects.checkFromToIndex(Objects.java:411)
        at java.base/java.lang.Integer.parseInt(Integer.java:708)
        at java.base/sun.net.www.ParseUtil.unescape(ParseUtil.java:166)
        at java.base/sun.net.www.ParseUtil.decode(ParseUtil.java:202)
        at java.base/sun.net.www.protocol.ftp.FtpURLConnection.<init>(FtpURLConnection.java:203)
        at java.base/sun.net.www.protocol.ftp.Handler.openConnection(Handler.java:61)
        at java.base/sun.net.www.protocol.ftp.Handler.openConnection(Handler.java:56)
        at java.base/java.net.URL.openConnection(URL.java:1094)
        at TestURL.main(TestURL.java:6)
```

tkiriyama commented:
> MalformedURLException should probably be thrown instead of simply accepting a malformed % encoded pair.

Thank you very much for your comment.  
I think we should accept any codes like "%25%s%G1" according to URL standard.  

Is there a reason why an exception should be thrown?

dfuch commented:
> I think we should accept any codes like "%25%s%G1" according to URL standard.
> 
> Is there a reason why an exception should be thrown?

Yes. It is not a valid escape sequence according to the spec.

jshell> URI.create("ftp://host:5678/%25%s%G1");
|  Exception java.lang.IllegalArgumentException: Malformed escape pair at index 19: ftp://host:5678/%25%s%G1
|        at URI.create (URI.java:906)
|        at (#1:1)
|  Caused by: java.net.URISyntaxException: Malformed escape pair at index 19: ftp://host:5678/%25%s%G1
|        at URI$Parser.fail (URI.java:2973)
|        at URI$Parser.scanEscape (URI.java:3101)
|        at URI$Parser.scan (URI.java:3124)
|        at URI$Parser.checkChars (URI.java:3142)
|        at URI$Parser.parseHierarchical (URI.java:3226)
|        at URI$Parser.parse (URI.java:3174)
|        at URI.<init> (URI.java:623)
|        at URI.create (URI.java:904)
|        ...

tkiriyama commented:
> It is not a valid escape sequence according to the spec.

I see.  JavaAPI document says that Escaped octets, that is, triplets consisting the percent character ('%') followed by two hexadecimal digits. Also, RFC 2396 says that  '%' must be escaped as "%25" in order to be used as data within a URI.
The exception thrown by URI.create() may be appropriate. I consider whether to fix the exception thrown by URL.openConnection().

tkiriyama commented:
I am very sorry for taking long time to fix.
I fixed java.net.URL#openConnection() to throw MalformedURLException when ParseUtil#decode() throws  IllegalArgumentException  because Javadoc says that openConnection() throws  IOException if if an I/O exception occurs.
Also I fixed  ParseUtil#decode() not to throw IndexOutOfBoundsException when a malformed escape pair is passed.

tkiriyama commented:
@dfuch 
Sorry to bother you, but I want you to review this fix again.

dfuch commented on src/java.base/share/classes/sun/net/www/ParseUtil.java:
```diff
@@ -200,8 +200,8 @@ public static String decode(String s) {
                     bb.put(unescape(s, i));
-                } catch (NumberFormatException e) {
-                    throw new IllegalArgumentException();
+                } catch (NumberFormatException | IndexOutOfBoundsException e) {
+                    throw new IllegalArgumentException("Malformed escape pair: " + s);
```
I would suggest turning the assert at line 200 into a proper bound check and throw IllegalArgumentException if `n - i >= 2`

dfuch replied:
I meant if `n - i` is **not** `>= 2` of course.

dfuch commented on src/java.base/share/classes/java/net/URL.java:
```diff
@@ -1092,7 +1092,13 @@ public URI toURI() throws URISyntaxException {
      */
     public URLConnection openConnection() throws java.io.IOException {
-        return handler.openConnection(this);
+        URLConnection url = null;
+        try {
+            url = handler.openConnection(this);
+        } catch (IllegalArgumentException e) {
+            throw new MalformedURLException(e.getMessage());
+        }
+        return url;
```
Please revert the changes in java.net.URL. It's not the proper place to do the conversion. If a conversion is needed, it should take place in the handler.
One possible experiment could be to  change `ParseUtil::decode` to declare `throws MalformedURLException` and see what no longer compiles. If compilation passes - then maybe that's the way to go. If it doesn't - then we could analyze what to do on a case-by-case basis.

tkiriyama replied:
There are some methods call ParseUtil::decode and of these 10 are not declared to be thrown MalformedURLException.
Modifying ParseUtil :: decode has not a little impact. Is modifying java.net.URL not proper?

The following classes have the methods which are not declared to be thrown MalformedURLException.

```
java.net.URLClassLoader
sun.net.www.protocol.ftp.FtpURLConnection
sun.security.provider.PolicyFile
jdk.internal.loader.FileURLMapper
```

tkiriyama replied:
@dfuch 
Could you tell me your opinion based on this information, please?

dfuch replied:
Let me try to summarize: there are several things that we could do:

 1. nothing: continue throwing IIOBE
 2. throw IAE instead of IIOBE, and let IAE trickle up the calling stack to the callers code
 3. throw MUE instead of IIOBE, and then update all the places where ParseUtils::decode is called to either let the MUE trickle up the stack if possible, or deal with it if not.
 4. throw IAE instead of IIOBE and somewhere up the stack catch that and transform it into MUE.
   4.1. do the catch & transform in URL::openConnection when calling URLStreamHandler::openConnection
   4.2. do the catch & transform in every protocol handler implementation that use ParseUtils::decode
    
Now let's see pro & cons:
 1. is not ideal because it looks like a bug (and it is)
 2. is expedient. There seem to be some precedent for that since ParseUtils::decode already throws IAE if it can't parse the port. However IAE is not specified to be thrown by openConnection. So maybe we should update the specification of URL::openConnection and URLStreamHandler::openConnection to allow for IAE. This is however a bit strange to use IAE because in this case the URL is indeed malformed, so MUE would be more appropriate.
 3. has the advantage of being clear, and force all callers of ParseUtils to deal with the situation. However, ParseUtils seems to be widely used, so each places where it's called needs to be examined, to see what would be the appropriate solution in each case 
 4. The advantage here is that callers of ParseUtils don't need to be changed if IAE is already appropriate for them. But those for which MUE would be more appropriate would beneficiate from having IAE transformed in MUE, which can be achieved by 4.1 or 4.2
  4.1 the issue here is that URLStreamHandler::openConnection is not specified to throw IAE, yet we catch IAE that might be thrown by custom URLStreamHandler implementation not present in the JDK, where the IAE that they throw might have nothing to do with a Malformed URL. So we risk introducing regressions for applications that provide their own custom URLStreamHandlers implementations, and that might not expect the wrapping of the IAE they throw into MUE.  I believe that if we did that it would need to be specified (in URL::openConnection) and it would require a CSR.
  4.2 we need to examine all protocol handler present in the JDK and do the wrapping there. URLStreamHandler::openConnection is specified to throw IOException, so throwing MUE is within the bounds. However - if we don't have ParseUtils::decode throw MUE, we're not guaranteed not to miss some places.
  
I'd say from a consistency point of view, 3. is probably best. Then failing that, from a compatibility vs consistency point of view 4.2 might be our best shot. 2. might be envisaged, but maybe at the expense of logging a follow up issue to clean that up and later decide between 3. or 4.2.

tkiriyama replied:
Thank you for your valuable comments. I think 3 is a good idea.  I will try fix, but it takes some time to see the impact on many other methods.

tkiriyama replied:
@dfuch 

I'm sorry, it took time for me to think about how to fix.
First, I thought of option 3, but it seems difficult.
In many of the methods in error, the javadoc defines the exceptions that can be thrown. 
When option 3, we will wrap exceptions in many methods because it is difficult to fix specifications of all methods and CSR.
However, if we are going to wrap exceptions, I think it's better and less fix to do it in a protocol handler, it means option 4.2.

Let me hear what you think.

dfuch replied:
4.2 might be fine especially if we can do it at the point were ParseUtils is called.

tkiriyama replied:
@dfuch 
Thank you very much. Your advice was very helpful in fix.
I fixed sun.net.www.protocol.ftp.Hanler.openConnection to throw MalformedURLException. 
Hanler classes in other protcol should not be fixed because they don't use ParseUtil.decode method.

dfuch commented on src/java.base/share/classes/sun/net/www/protocol/ftp/Handler.java:
```diff
@@ -58,6 +59,12 @@ protected java.net.URLConnection openConnection(URL u)
+        try {
+            connection = new FtpURLConnection(u, p);
+        } catch (IllegalArgumentException e) {
+            throw new MalformedURLException(e.getMessage());
```
I'd suggest keeping the IllegalArgumentException as the root cause of the MUE for better diagnosability. Otherwise LGTM.

tkiriyama replied:
The constructor "MalformedURLException (Throwable cause)" doens'nt exist. Isn't there any other method?

dfuch replied:
You can use `Throwable::initCause`

dfuch replied:
```
var mfue = new MalformedURLException(e.getMessage());
mfue.initCause(e);
throw mfue;
```

tkiriyama replied:
Thank you for your advice, I could fix it.

dfuch commented:
I'm going to run some tests in our CI and will come back here with the result.

dfuch commented on src/java.base/share/classes/sun/net/www/ParseUtil.java:
```diff
@@ -197,11 +197,14 @@ public static String decode(String s) {
             int ui = i;
             for (;;) {
-                assert (n - i >= 2);
+                if (n - i >= 2) {
```
OK - the test should be inverted - sorry for not noticing that earlier. It should be `n - i < 2`. It's causing massive failures in the CI. Can you please run at least the tests under jdk/sun/net and jdk/java/net after making the change?

tkiriyama replied:
I am sorry for my bad mistake.
I fixed it and run the tests under jdk/sun/net and jdk/java/net. I confirmed all tests passed in Linux.

dfuch commented:
Thanks! I will rerun the tests and if everything comes back green I'll approve.

dfuch reviewed (approved):
Tests were green: approved.

tkiriyama commented:
@dfuch 
Thank you very much for your review and some helpful comments.

I think two reviews are required to be integrated. Could anyone please review this pull request?

dfuch commented:
For core libraries one reviewer is usually enough. The two reviewers rule applies mostly to hotspot components I believe. You may wait for a second reviewer if you wish to, though, or if the first reviewer advises you to do so.

tkiriyama commented:
@dfuch 
Thank you, Your advice has been really helpful.
I had already gotten your valuable review, so I would be glad if you could sponsor this PR.

tkiriyama commented:
/integrate

dfuch commented:
/sponsor

dfuch commented:
Done! Thanks for your perseverance!
